### PR TITLE
Cram: sanitize error messages

### LINKED ISSUE #3673: Cram: error messages when redirecting IO
Hi,

When there is an error with redirecting input or output, the message leaks some info about the temporary shell script:

```diff
 Input:
 
   $ <doesnotexist cat
+  /tmp/dune.cram.ee5255run.t/main.sh: 1: /tmp/dune.cram.ee5255run.t/14.sh: cannot open doesnotexist: No such file
+  [2]
 
 Output:
 
   $ echo > does/not/exist
+  /tmp/dune.cram.ee5255run.t/main.sh: 1: /tmp/dune.cram.ee5255run.t/16.sh: cannot create does/not/exist: Directory nonexistent
+  [2]
```

I'm not sure what the expected behavior is. Maybe just rewrite the beginning so that it reads "Internal dune-cram error: cannot open doesnotexist: No such file" ?

(cc @rgrinberg ; I found this while testing the cram stanza)

Thanks!

### PR DESCRIPTION
**Pull request by rgrinberg**
Fix #3673

### CODE REVIEW

**Comment by rgrinberg** [comment]
Hmm, I'm unable to reproduce the error in CI. @emillon mind trying this out? (if you use Linux)

File: src/dune/cram_exec.ml [hunk 1]
```diff
@@ -74,7 +80,7 @@ end = struct
         Code_error.raise "unable to parse csexp" [ ("error", String error) ]
       | Ok None -> ()
       | Ok (Some sexp) ->
-        let command = Command.of_sexp sexp in
+        let command = Command.of_sexp (assert false) sexp in
```
Comment (rgrinberg) [inline src/dune/cram_exec.ml:83]: This is dead code anyway

**Comment by jeremiedimino** [comment]
The printing with `git diff` in the CI doesn't look right BTW.

**Comment by rgrinberg** [comment]
> The printing with git diff in the CI doesn't look right BTW.

Unfortunately, I'm unable to reproduce the issue locally. I assume this is a difference of behavior in the shells that are being used. Do you think it's worth delaying the release for this? I understand that it would be better to have the same behavior since the beginning, but this doesn't seem possible in this case if we're willing to keep updating the regular expressions for different shells.

**Comment by jeremiedimino** [comment]
That seems fine. We'll just need to tie the regexp to the version of the dune language, which seems fine.

Regarding this PR, when I look at the CI failure, there seems to be two issues:
- there is still a `: ` prefix (I can reproduce it on my machine)
- dune calls `git diff --no-index ...` as diff program and instead of printing a diff between `run.t` and `run.t.corrected`, it shows a diff between `run.t` and `/dev/null` and then a diff between `/dev/null` and `run.t.corrected`

The second problem seems independent of this PR, but it'd be good to look into it at some point. When patdiff is not installed, dune calls `git diff` by default and this double diffing behaviour is a pretty bad user experience.

**Comment by rgrinberg** [comment]
> there is still a : prefix (I can reproduce it on my machine)

I believe I fixed that.

> dune calls git diff --no-index ... as diff program and instead of printing a diff between run.t and run.t.corrected, it shows a diff between run.t and /dev/null and then a diff between /dev/null and run.t.corrected

That's not exactly what's happening. The diff command seems to be correct, but there seems to be a weird interaction between promotion, sandboxing, and git's diff that's causing the issue. AFAIK, the way promotion works is as follows:

* We run diff (git diff in this case) against the two files.
* If diff detects a difference, we move the corrected file into the promotion staging area.
* We register the file from the staging area to be promoted.

As far as I can tell, there seems to be some sort of a race and `git diff` is observing the file after it was moved to the staging area. I have no idea why it's happening because `git diff` should finish running before we move the file to staging area, but this is the problem from what I've diagnosed.

In any case, this is indeed unrelated to this PR and I've tweaked the error message to match Linux. I've disabled the test on MacOS as I'm getting a different error message.

**Comment by jeremiedimino** [comment]
The regexp should be anchored, but otherwise looks good to merge.

File: src/dune/cram_exec.ml [hunk 2]
```diff
@@ -136,11 +142,12 @@ let run_expect_test file ~f =
 
 let _BUILD_PATH_PREFIX_MAP = "BUILD_PATH_PREFIX_MAP"
 
-let extend_build_path_prefix_map ~env ~cwd =
+let extend_build_path_prefix_map ~env ~cwd ~temp_dir =
   let s =
     Build_path_prefix_map.encode_map
       [ Some
           { source = Path.to_absolute_filename cwd; target = "$TESTCASE_ROOT" }
+      ; Some { source = Path.to_absolute_filename temp_dir; target = "$TMPDIR" }
```
Comment (rgrinberg) [inline src/dune/cram_exec.ml:150]: document this variable being sanitized.

**Pull request merged**